schema/productconfig.rnc

```diff
@@ -0,0 +1,475 @@
+# RELAX NG Schema for Docserv² product configuration
+
+# BASICS
+
+default namespace = ""
+
+# FIXME: Should this be <docservconfig> instead? (I guess it makes more sense
+# to be able to validate individual productconfig files instead of validating
+# an essentially internal representation)
+start = ds.productindex
+
+
+# TYPES
+
+ds.type.class =
+  ## A class or space-separated list of classes
+  xsd:token { pattern = "[\-_\.a-zA-Z0-9 ]*" }
+
+ds.type.email =
+  # http://www.dpawson.co.uk/relaxng/schema/datatypes/datatypes.html#d1034e184
+  # Works for both normal addresses and forms like
+  # "David Tolpin"@[Obscure Place]
+  # The expression above is slightly more allowing than required,
+  # but should be appropriate for the majority of cases.
+  xsd:token {
+    pattern="""([a-zA-Z0-9!#$%&'*+\-/=?\^_`{|}~]+(\.[a-zA-Z0-9!#$%&'*+\-/=?\^_`{|}~]+)*|"[^"\\]*")@([a-zA-Z0-9!#$%&'*+\-/=?\^_`{|}~]+(\.[a-zA-Z0-9!#$%&'*+\-/=?\^_`{|}~]+)*|\[[^\[\]\\]*\])"""
+  }
+
+ds.type.lang =
+  # A language code
+  xsd:token {
+    pattern="""[a-z]{2,5}(-[a-z]{2,5})?"""
+  }
+
+ds.type.gitremote =
+  # URL to a Git remote
+  # (Let's not support SSH URLs for the moment since that needs authentication)
+  xsd:token {
+    pattern="""https?://.*"""
+  }
+
+ds.type.dirname =
+  # A relative directory name
+  xsd:token {
+    pattern="""[\-_\.\+a-zA-Z0-9][\-_\.\+/a-zA-Z0-9]*"""
+  }
+
+ds.type.branch =
+  # A Git branch name
+  xsd:token {
+    pattern="""[\-_\.\+/a-zA-Z0-9]+"""
+  }
+
+ds.type.dcfile =
+  # A DC file name
+  xsd:token {
+    pattern="""DC-[\-_\.\+/a-zA-Z0-9]+"""
```

Comment:
Maybe you should restrict only the start string ("DC-"), but not how a DC file is composed of which characters. Probably `xsd:anyURI` would work here too.

Reply:
DAPS itself fails on so many file names (because GNU make a tool of this century not is), I think I'd rather keep the restrictions here. I used xsd:anyURI though.